omniduct/databases/cursor_formatters.py

```diff
@@ -0,0 +1,146 @@
+import csv
+import io
+
+
+class CursorFormatter(object):
+
+    def __init__(self, db_client, cursor, **kwargs):
+        self.db_client = db_client
+        self.cursor = cursor
+        self.init(**kwargs)
+
+    def init(self):
+        pass
+
+    @property
+    def column_names(self):
+        return [c[0] for c in self.cursor.description]
+
+    @property
+    def column_formats(self):
+        return [c[1] for c in self.cursor.description]
+
+    def dump(self):
+        try:
+            data = self.cursor.fetchall()
+            out = self.format_dump(data)
+        finally:
+            self.cursor.close()
+        return out
+
+    def stream(self, batch=None):
+        try:
+            column_names = self.column_names
+            column_formats = self.column_formats
+
+            if batch is not None:
+                while True:
+                    b = self.cursor.fetchmany(batch)
+                    if len(b) == 0:
+                        return
+                    yield self.format_dump(b)
+            else:
+                for row in self.cursor:
+                    yield self.format_row(row)
+        finally:
+            self.cursor.close()
+
+    def format_dump(self, data):
+        raise NotImplementedError("{} does not support formatting dumped data.".format(self.__class__.__name__))
+
+    def format_row(self, row):
+        raise NotImplementedError("{} does not support formatting streaming data.".format(self.__class__.__name__))
+
+
+class PandasCursorFormatter(CursorFormatter):
+
+    def init(self, index_fields=None, date_fields=None):
+        self.index_fields = index_fields
+        self.date_fields = date_fields
+
+    def format_dump(self, data):
+        import pandas as pd
+
+        df = pd.DataFrame(data=data, columns=self.column_names)
+
+        if self.date_fields is not None:
+            try:
+                df = pandas.io.sql._parse_date_columns(df, self.date_fields)
+            except:
+                logger.warning('Unable to parse date columns. Perhaps your version of pandas is outdated.')
+
+        if self.index_fields is not None:
+            df.set_index(self.index_fields, inplace=True)
+
+        return df
+
+    def format_row(self, row):
+        import pandas as pd
+
+        # TODO: Handle parsing of date fields
+
+        return pd.Series(row, index=self.column_names)
+
+
+class DictCursorFormatter(CursorFormatter):
+
+    def format_dump(self, data):
```

Comment:
I like the abstract base class, since it allows subclasses to partially implement the API, which may make sense in some cases. When it does make sense, appropriate error messages will automatically be generated. Also... I'm not sure there's a good sense of that a "default" implementation should look like. So I'll leave it as is for now :).